### PR TITLE
Use lat/lon for Open Weather Map

### LINKED ISSUE #1485: Always use "lat, lon" for OpenWeatherMap requests
Using anything other than `lat, lon` for the current weather API is deprecated. Also, city name and zip-code can both be rather inaccurate (which affects `autolocate`).

From https://openweathermap.org/current :
> We highly recommend you using our [Geocoding API](https://openweathermap.org/api/geocoding-api). This API uses a much more accurate geographical database and supports both direct and reverse geocoding.
> 
> Please note that built-in API requests by city name, zip-codes and city id will be deprecated soon.

Deprecating `city id` does seem a little strange, though.

I suggest making the following changes:

- [ ] Keep using ipapi.co for `autolocate`, but use the coordinates instead of city name (#1593).
- [ ] Use the OMW geocoding API to convert `place` into `lat, lon`. Add a note in docs that this will consume API calls. Also, make it clear the format is `city, country code`.
- [ ] Add a `zip` key. Again, this consumes an API call.
- [ ] Keep `city_id` for now, but add a note that it may stop working in the future.

### PR DESCRIPTION
**Pull request by bim9262**
Fixes #1485

### CODE REVIEW

**Comment by MaxVerevkin** [comment]
Do you think it makes sense to remove useless spaces/newlines in `open_weather_map_cities.json` like `met_no_legends.json`?

**Comment by bim9262** [comment]
If there's ever an update to the files it would be very hard to read the diff (as a sanity check) if it's all on a single line. Since the line breaks are not a factor binary size when using bincode I'd leave the line breaks there. Also where did you find the met_no_legends? I can add it to the update script.

**Comment by bim9262** [comment]
I'm considering restructuring the json file to keep the keep the id as a number so that we can pack it smaller by not having strings as keys (need to do some testing though)

**Comment by MaxVerevkin** [comment]
> If there's ever an update to the files it would be very hard to read the diff (as a sanity check) if it's all on a single line

Good point.

> Also where did you find the met_no_legends?

Here: https://web.archive.org/web/20230328155110/https://api.met.no/weatherapi/weathericon/2.0/legends.

**Comment by MaxVerevkin** [comment]
> to keep the keep the id as a number

Yeah, I just wanted to ask why you converted them to strings, as it doesn't work is some cases. For example the last entry of the file is

```json
  "7.12969E+7": {
    "lon": 33.92,
    "lat": 44.59
  }
```

**Comment by MaxVerevkin** [comment]
Is it necessary to prefetch this info? In case with met.no we had to because they disabled the API. The 1M binary increase is unfortunate.

**Comment by bim9262** [comment]
There's two errors like that, but they come from the source file. I opened a support case with OWM...

Based on the errors in the other city list they publish they accidentally made them floats (those same IDs ended with .0 in that format)

**Comment by bim9262** [comment]
> Is it necessary to prefetch this info? In case with met.no we had to because they disabled the API. The 1M binary increase is unfortunate.

We could fetch it on demand. We'd probably want to give the lowest precedence to city_id since it would become the most expensive to load.

File: src/blocks/weather/open_weather_map.rs [hunk 1]
```diff
@@ -79,24 +95,184 @@ struct ApiWeather {
...
+                        .map(|sys_time| {
+                            DateTime::<Utc>::from(sys_time)
+                                .format("%a, %d %b %Y %H:%M:%S GMT")
+                                .to_string()
+                        });
+
+                    let cities_request = if let Ok(timestamp) = cities_temp_timestamp.as_ref() {
+                        debug!("{}", timestamp);
+                        REQWEST_CLIENT.get(CITIES_URL).header(
+                            reqwest::header::IF_MODIFIED_SINCE,
```
Comment (bim9262) [inline src/blocks/weather/open_weather_map.rs:200]: This isn't being respected by owm's server. Curl is smart enough to handle it, but reqwest/hyper doesn't yet.

**Comment by bim9262** [comment]
This is now with live fetching. It caches the result of the file to disk and compares the last modified date on the server with the local modification date. (owm does not respect the `If-Modified-Since` header).

Switched over to the larger city list (as described at https://openweathermap.org/bulk#list):
List of major cities across the globe - 22,635 cities. You can download the full list of cities [here](https://bulk.openweathermap.org/sample/current.city.list.json.gz)
Extended cities list across the globe - 209,579 cities. You can download the full list of cities [here](https://bulk.openweathermap.org/sample/city.list.json.gz)

File: src/blocks/weather/open_weather_map.rs [hunk 2]
```diff
@@ -79,24 +94,198 @@ struct ApiWeather {
...
+    id: f32,
+    coord: CityCoord,
+}
+#[derive(Deserialize, Debug)]
+struct CityCoord {
+    lat: f64,
+    lon: f64,
+}
+
+static LOCATION_CACHE: Mutex<Option<String>> = Mutex::new(None);
```
Comment (MaxVerevkin) [inline src/blocks/weather/open_weather_map.rs:109]: Does this prevent from having two blocks which show weather in different places?

File: src/blocks/weather/open_weather_map.rs [hunk 3]
```diff
@@ -79,24 +94,198 @@ struct ApiWeather {
...
+            None => match self.config.city_id.as_ref() {
+                Some(city_id) => {
+                    let city_id = &city_id.parse().error("city_id must be numeric")?;
+
+                    let mut cities_temp_path = env::temp_dir();
+                    let user = env::var("USER");
+                    cities_temp_path.push(format!(
+                        "city.list.min.json.gz-{}",
+                        user.as_deref().unwrap_or("no-user")
+                    ));
```
Comment (MaxVerevkin) [inline src/blocks/weather/open_weather_map.rs:208]: Perhaps it's better to use `dirs::cache_dir()` instead of temp?

Comment (MaxVerevkin) [reply src/blocks/weather/open_weather_map.rs:208]: By the way, maybe extract this cache and fetching logic into a different function?

File: src/blocks/weather/open_weather_map.rs (hunk 2 above)
Comment (bim9262) [reply src/blocks/weather/open_weather_map.rs:109]: Yes... Good catch.

File: src/blocks/weather.rs [hunk 4]
```diff
@@ -105,7 +107,7 @@ pub mod open_weather_map;
 
 const IP_API_URL: &str = "https://ipapi.co/json";
 
-static LAST_AUTOLOCATE: Mutex<Option<AutolocateResult>> = Mutex::new(None);
+static LAST_AUTOLOCATE: Lazy<Mutex<Option<AutolocateResult>>> = Lazy::new(|| Mutex::new(None));
```
Comment (MaxVerevkin) [inline src/blocks/weather.rs:110]: Why this change? This mutex is not locked across `.await`s. In cases like this [tokio advises to use ordinary mutexes](https://docs.rs/tokio/latest/tokio/sync/struct.Mutex.html#which-kind-of-mutex-should-you-use).

File: src/blocks/weather/open_weather_map.rs [hunk 5]
```diff
@@ -79,24 +94,233 @@ struct ApiWeather {
...
+#[derive(Eq, Hash, PartialEq, Debug)]
+enum LocationKeys {
+    CityId(String),
+    Place(String),
+    Zip(String),
+}
 
-        let location_query = autolocated
+static LOCATION_CACHE: Lazy<Mutex<HashMap<LocationKeys, String>>> =
+    Lazy::new(|| Mutex::new(HashMap::new()));
```
Comment (MaxVerevkin) [inline src/blocks/weather/open_weather_map.rs:117]: Why can't we just store the cached location string in the `Service` struct?

**Comment by MaxVerevkin** [comment]
> we don't update the list or correct issues due to it being mainly open source

I can't make sense of this sentence...

> So perhaps we should continue to send the city_id, but query the geo api for place and zip.

If they won't actually remove support for this, and it makes our code simpler, then why not.

**Comment by bim9262** [comment]
> > we don't update the list or correct issues due to it being mainly open source
> 
> I can't make sense of this sentence...
> 
> > So perhaps we should continue to send the city_id, but query the geo api for place and zip.

Keep the code I wrote for the place name and zip codes in this pr, but use the city_id like we were doing originally

File: src/blocks/weather.rs (hunk 4 above)
Comment (bim9262) [reply src/blocks/weather.rs:110]: I did that because of the use super::* in own, but can use the two mutex types with an alias or qualify wich mutex to use

File: src/blocks/weather/open_weather_map.rs (hunk 5 above)
Comment (bim9262) [reply src/blocks/weather/open_weather_map.rs:117]: I tried that initially, but could not figure out how to fix the error caused when you make the service mutable. The line above the retry becomes a FnMut, and complains: `Captured variable cannot escape FnMut closure body`

**Comment by MaxVerevkin** [comment]
> Keep the code I wrote for the place name and zip codes in this pr, but use the city_id like we were doing originally

I was referring to their rationale for not fixing city IDs. "open source" == "unmaintained"? But yes, let's use city_id as we were doing it.

File: src/blocks/weather/open_weather_map.rs (hunk 5 above)
Comment (MaxVerevkin) [reply src/blocks/weather/open_weather_map.rs:117]: Oh, yeah, that's an annoying error. Can we construct the location string in `Service::new()` then? It can take `autolocate: bool` and do the necessary stuff if it is `false`.

File: src/blocks/weather.rs (hunk 4 above)
Comment (MaxVerevkin) [reply src/blocks/weather.rs:110]: If it is possible to get rid of `LOCATION_CACHE` then this will resolve itself, I guess.

**Review by MaxVerevkin — APPROVED** [review]
Looks good

**Comment by bim9262** [comment]
> > Keep the code I wrote for the place name and zip codes in this pr, but use the city_id like we were doing originally
> 
> I was referring to their rationale for not fixing city IDs. "open source" == "unmaintained"? But yes, let's use city_id as we were doing it.

I was half asleep when I read your message. Yeah I don't understand  "open source" == "unmaintained" either.

This was the response about where the data comes form:

>We get our city database from https://www.geonames.org/. However, you can't find the weather IDs there, IDs were generated by us. Since the list has been deprecated and was not updated for a while some city IDs could be missed or point to other coordinates.
>
>We strongly recommend using geographical coordinates instead of city IDs and using Geocoder API: https://openweathermap.org/api/geocoding-api
